riscv/src/compiler.rs

```diff
@@ -635,7 +635,78 @@ fn preamble(degree: u64, coprocessors: &CoProcessors, with_bootloader: bool) ->
 "#
 }
 
-fn memory() -> String {
+fn memory(with_bootloader: bool) -> String {
+    // There are subtle differences between the memory machines with and without continuations:
+    // - There is an extra `mstore_bootloader` instruction. For the most part, it behaves just
+    //   like `mstore`.
+    // - When `m_change` is true, the `m_is_bootloader_write` has to be true in the next row.
+    // - The `(1 - m_is_write') * m_change * m_value' = 0` constraint is removed, as we no longer can
+    //   have a read as the first operation on a new address.
+    // - The `(1 - m_change) * LAST = 0` constraint is replaced with
+    //   `LAST * (1 - m_change) * (m_addr + 1) = 0`. This allows for a valid assignment in the case
+    //   where there is no memory operation in the entire chunk: The address can be set to -1 (which
+    //   cannot be represented in 32 bits, hence there is can't be an actual memory operation
+    //   associated with it). In that case, `m_change` can be 0 everywhere.
+    let bootloader_specific_parts = if with_bootloader {
+        r#"
+    // Memory operation flags
+    col witness m_is_write;
+    col witness m_is_bootloader_write;
+    col witness m_is_read;
+
+    // All operation flags are boolean and either all 0 or exactly 1 is set.
+    m_is_write * (1 - m_is_write) = 0;
+    m_is_read * (1 - m_is_read) = 0;
+    m_is_bootloader_write * (1 - m_is_bootloader_write) = 0;
+    m_is_read * m_is_write = 0;
+    m_is_read * m_is_bootloader_write = 0;
+    m_is_bootloader_write * m_is_write = 0;
+
+    // The first operation of a new address has to be a bootloader write
+    m_change * (1 - m_is_bootloader_write') = 0;
```

Comment:
In that case, won't it lead to 1=0?

Reply:
In what case? If `m_change` is 1, `m_is_bootloader_write` has to be 1 in the next row, so then it's `0 = 0`.

Reply:
true actually